### PR TITLE
Autocomplete: truncate multiline completions based on parse-trees

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

- Truncate Typescript and Javascript multi-line completions based on tree-sitter parse-trees:
  - Paste the received suggestion into the document.
  - Parse it with tree-sitter.
  - Query the first block-like statement from the range of the inserted code. 
  - Use the received node to truncate the completion. 
  - This logic is still behind the experimental setting `autocompleteExperimentalSyntacticPostProcessing`.
- Part of #910 
- Planned follow-up https://github.com/sourcegraph/cody/issues/1056
- Based on https://github.com/sourcegraph/cody/pull/938 and https://github.com/sourcegraph/cody/pull/947

## Test plan

Added unit tests + all of the existing multiline-truncation tests are applicable.

### CODE REVIEW

File: vscode/src/completions/text-processing/truncate-parsed-completion.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ */
+export function truncateParsedCompletion(context: CompletionContext): string {
+    const { completion, document } = context
+
+    const parseTreeCache = getCachedParseTreeForDocument(document)
+
+    // Do nothig if the syntactic post-processing is not enabled.
+    if (!completion.tree || !parseTreeCache) {
+        return completion.insertText
+    }
```
Comment (philipp-spiess) [inline vscode/src/completions/text-processing/truncate-parsed-completion.ts:51]: Should we through here instead, that seems like a contract violation with how we except this function to be called

**Review by philipp-spiess — APPROVED** [review]
Good stuff! Can you add logging to analytics as to when this code path is used?

**Comment by valerybugakov** [comment]
> Can you add logging to analytics as to when this code path is used?

Yes, will do it in a follow-up https://github.com/sourcegraph/cody/issues/1070

File: vscode/src/completions/text-processing/truncate-parsed-completion.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import { ParsedCompletion } from './parse-completion'
+
+interface CompletionContext {
+    completion: ParsedCompletion
+    document: TextDocument
+}
+
+export const MULTILINE_TRUNCATION_SUPPORTED_LANGUAGES: Set<string> = new Set([
+    SupportedLanguage.JavaScript,
+    SupportedLanguage.TypeScript,
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply vscode/src/completions/text-processing/truncate-parsed-completion.ts:15]: Added them to the list 👍

**Pull request merged**